src/lib/feed.ts

```diff
@@ -43,97 +19,57 @@ export interface RankLinksOptions {
-
-  if (cached) {
-    console.log("CACHE HIT", cacheKey);
-    items = JSON.parse(cached.value);
-  } else {
-    console.log("CACHE MISS", cacheKey);
-    const minTime = `DATETIME('${startTime}', '-${range}')`;
-    const maxTime = `DATETIME('${startTime}')`;
-
-    const posts = db
-      .prepare(
-        `
-          SELECT 
-              post.did,
-              post.rkey,
-              post.url,
-              post.createdAt,
-              COUNT(CASE WHEN reaction.type = 'like' THEN 1 END) AS likes,
-              COUNT(CASE WHEN reaction.type = 'repost' THEN 1 END) AS reposts,
-              COUNT(CASE WHEN reaction.type = 'comment' THEN 1 END) AS comments
-          FROM 
-              post
-              LEFT JOIN reaction ON post.rkey = reaction.rkey AND post.did = reaction.did
+  const cursorTime = cursor.time || "now";
+  const cursorDateTime = `DATETIME('${cursorTime}', '-${range}')`;
+
+  const posts = cacheDb
+    .prepare(
+      `
+      SELECT
+        did,
+        rkey,
+        url,
+        createdAt,
+        score,
+        dateWritten
+      FROM
+        post
+      WHERE
+        createdAt >= ${cursorDateTime} AND
+        dateWritten = (
+          SELECT
+            dateWritten
+          FROM
+            date_written
           WHERE
-              post.createdAt >= ${minTime}
-              AND post.createdAt <= ${maxTime}
-              AND (reaction.createdAt >= ${minTime} AND post.createdAt <= ${maxTime} OR reaction.createdAt IS NULL)
-          GROUP BY
-              post.did, post.rkey, post.url, post.createdAt;
+            dateWritten >= ${cursorDateTime}
+          ORDER BY
+            julianday(dateWritten) - julianday('${cursorTime}')
+          DESC
+          LIMIT 1
+        )
+      ORDER BY
+        score DESC
+      LIMIT
+        ${limit}
+        ${cursor.index ? `OFFSET ${cursor.index}` : ""};
         `
-      )
-      .all() as PostWithData[];
-
-    const top: Record<string, PostWithData[]> = {};
-
-    for (let i = 0; i < posts.length; i++) {
-      const post = posts[i];
-
-      if (!post) {
-        continue;
-      }
-
-      if (!top[post.url]) {
-        top[post.url] = [];
-      }
-
-      top[post.url]?.push(post);
-    }
-
-    items = Object.entries(top)
-      .sort((a, b) => sumPostScore(b[1]) - sumPostScore(a[1]))
-      .map(([url, posts]): [string, PostWithData[]] => [
-        url,
-        posts.sort((a, b) => getPostScore(b) - getPostScore(a)),
-      ])
-      .slice(0, 1500);
-
-    db.prepare(`INSERT OR REPLACE INTO cache (key, value) VALUES (?, ?)`).run(
-      cacheKey,
-      JSON.stringify(items)
-    );
-  }
-
-  if (cursor.index) {
-    items = items.slice(cursor.index + 1);
-  }
-
-  items = items.slice(0, limit);
+    )
+    .all() as PostWithData[];
 
   return {
-    items: Object.fromEntries(items),
-    cursor: `${startTime}/${(cursor.index || 0) + limit}`,
+    items: posts,
+    cursor: `${cursor.time || posts[posts.length - 1]?.dateWritten}/${
```

Comment:
I made 2 changes. I noticed that when requesting a feed with a cursor with would return a newer time. Now it should always re-use the cursor and preserve the feed the user is looking at

Reply:
If the cursor time is not from the stored write times, we give you the closest newest view that is cached. By keeping the time that isn't from the cache we end up doing that lookup each time (this is a minor performance impact) but ultimately they are getting the same view either way.

This is only a case if it is an old cursor before my changes, or if a cached view happened to be removed after a while when viewing (at which point whatever they were looking at before would have changed under their feet anyways if that makes sense).

It is a minor thing overall as either way they should get the same view of the feed so long as the cursor time isn't changing between pulls (other than in the case we set the cursor time to the one tied to the view they are getting).

I had the code this way eventually, but it seemed more correct to give back the actually viewed written date which will absolutely guarantee to view the same view.